aspnetcore/blazor/test.md

```diff
@@ -0,0 +1,132 @@
+
+## Test components with bUnit
+
+There's no official Microsoft testing framework for Blazor, but the community-driven project [bUnit](https://github.com/egil/bUnit) provides a convenient way to unit test Blazor components.
+
+> [!NOTE]
+> bUnit is a third-party testing library and isn't supported or maintained by Microsoft or the .NET Foundation.
+
+bUnit works with general-purpose testing frameworks, such as [MSTest](/dotnet/core/testing/unit-testing-with-mstest), [NUnit](https://nunit.org/), and [xUnit](https://xunit.github.io/). These testing frameworks make bUnit tests look and feel like regular unit tests. bUnit tests integrated with a general-purpose testing framework are ordinarily executed with:
+
+* [Visual Studio's Test Explorer](/visualstudio/test/run-unit-tests-with-test-explorer).
+* [`dotnet test`](/dotnet/core/tools/dotnet-test) CLI command in a command shell.
+* An automated DevOps testing pipeline.
+
+> [!NOTE]
+> Test concepts and test implementations across different test frameworks are similar but not identical. Refer to the test framework's documentation for guidance.
+
+The following demonstrates the structure of a bUnit test on the `Counter` component in an app based on a Blazor project template. The `Counter` component displays and increments a counter based on the user selecting a button in the page:
+
+```razor
+@page "/counter"
+
+<h1>Counter</h1>
+
+<p>Current count: @currentCount</p>
+
+<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
+
+@code {
+    private int currentCount = 0;
+
+    private void IncrementCount()
+    {
+        currentCount++;
+    }
+}
+```
+
+The following bUnit test verifies that the CUT's counter is incremented correctly when the button is selected:
+
+```csharp
+[Fact]
+public void CounterShouldIncrementWhenSelected()
+{
+    // Arrange
+    using var cxt = new TestContext();
+    var cut = ctx.RenderComponent<Counter>();
+    var paraElm = cut.Find("p");
+
+    // Act
+    cut.Find("button").Click();
+    var paraElmText = paraElm.TextContent;
+
+    // Assert
+    paraElmText.MarkupMatches("Current count: 1");
+}
+```
+
+The following actions take place at each step of the test:
+
+* *Arrange*: The `Counter` component is rendered using bUnit's `TestContext`. The CUT's paragraph element (`<p>`) is found and assigned to `paraElm`.
+
+* *Act*: The button's element (`<button>`) is located and then selected by calling `Click`, which should increment the counter and update the content of the paragraph tag (`<p>`). The paragraph element text content is obtained by calling `TextContent`.
+
+* *Assert*: `MarkupMatches` is called on the text content to verify that it matches the expected string, which is `Current count: 1`.
```

Comment:
```suggestion
* *Assert*: `MarkupMatches` is called on the text content to verify that it matches the expected string, which is `Current count: 1`.

> [!NOTE]
> The `MarkupMatches` assert method differs from a regular string comparison assertion, e.g. `Assert.Equal("Current count: 1", paraElmText);`, by performing a semantic comparison of the input and expected HTML markup. A semantic comparison is aware of HTML semantics, meaning things like insignificant whitespace is ignored, leading to much more stable tests. Learn more at https://bunit.egilhansen.com/docs/verification/semantic-html-comparison
```

Readers might wonder about `MarkupMatches`, so I think it is a good idea to give some insights into what and why. It could look something like this.